datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -1084,4 +971,209 @@ where
+            .collect::<Vec<_>>();
+
+        // Return a stream that lazily loads batches of blocks
+        Box::new(
+            stream::iter_ok::<_, Error>(
+                block_hashes_batches
+                    .into_iter()
+                    .map(|block_hashes_batch| {
+                        debug!(
+                            logger,
+                            "Requesting {} block(s) in parallel",
+                            block_hashes_batch.len()
+                        );
+
+                        // Start loading all blocks in this batch
+                        stream::futures_ordered(
+                            block_hashes_batch
+                                .into_iter()
+                                .map(|block_hash| self.load_block(&logger, block_hash)),
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            )
+            .flatten()
+            .collect()
+            .map(move |new_blocks| {
+                if let Err(e) = chain_store.upsert_thin_blocks(new_blocks.clone()) {
+                    error!(logger, "Error writing to block cache {}", e);
+                }
+                blocks.extend(new_blocks);
+                blocks.sort_by_key(|block| block.number);
+                stream::iter_ok(blocks)
+            })
+            .flatten_stream(),
+        )
+    }
+
+    fn block_range_to_ptrs(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+    ) -> Box<dyn Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send> {
+        let eth = self.clone();
+        let logger = logger.clone();
+
+        // Generate `EthereumBlockPointers` from `to` backwards to `from`
+        Box::new(
+            self.block_hash_by_block_number(&logger, to)
+                .map(move |block_hash_opt| EthereumBlockPointer {
+                    hash: block_hash_opt.unwrap(),
+                    number: to,
+                })
+                .and_then(move |block_pointer| {
+                    stream::unfold(block_pointer, move |descendant_block_pointer| {
+                        if descendant_block_pointer.number < from {
+                            return None;
+                        }
+                        // Populate the parent block pointer
+                        Some(
+                            eth.block_parent_hash(&logger, descendant_block_pointer.hash)
+                                .map(move |block_hash_opt| {
+                                    let parent_block_pointer = EthereumBlockPointer {
+                                        hash: block_hash_opt.unwrap(),
+                                        number: descendant_block_pointer.number - 1,
+                                    };
+                                    (descendant_block_pointer, parent_block_pointer)
+                                }),
+                        )
+                    })
+                    .collect()
```

Comment:
How large are the block ranges we'd pass in here? While simple and easy to follow, populating the vector one item at a time could be slow — a more performant (but more complicated) approach would be to optimistically resolve all hashes for the `(to, from)` range in parallel (or in parallel batches) and checking them for consistency afterwards.

Reply:
I've improved this to fetch the block hashes in parallel, and refactored the code that does regular block fetching so that they're similar. I also removed the special case for the scan range of block triggers, this helps a subgraph like Betoken to be less slow, though the situation is still not ideal because we don't cache the number -> hash association in the database. Consistency checking is not necessary since this range is for final blocks.